# Computational Methods

Describe the computational methods of the article on glycosylation and transcription on chromatin in mammalian cells. The description of the methodology and the code for each figure are available in the corresponding folders.

## Data

The data produced for this study can be found on ArrayExpress:

| Accession ID | Title |
|--------------|-------|
| [E-MTAB-14307](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14307) | CutnRun directed against O-GlcNac in the human DLD-1 cell line with doxycycline treatment to induce RNA Polymerase II degradation |
| [E-MTAB-14308](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14308) | CutnRun of O-GlcNac in mouse ES cells in high glucose condition |
| [E-MTAB-14309](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14309) | CutnRun of O-GlcNac in mouse ES cells in low glucose condition at 6 hours, 8 hours, and 1 week |
| [E-MTAB-14310](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14310) | ChIP-seq O-GlcNac in mouse ES cells |
| [E-MTAB-14313](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14313) | Total RNA-seq upon degradation of OGT by siRNA |
| [E-MTAB-14350](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14350) | ATAC-seq in the differentiation system ESC to NPC upon O-GlcNAc removal |

Processed data were also deposited on Zenodo: [DOI 10.5281/zenodo.12793185](https://zenodo.org/records/12793186)

All codes and conda environments are provided along with the figures in the corresponding folders.

The complete list of data used for the article can be found [here](alldata.csv)


## Acknowledgements

Data management, HPC nodes, Gitlab, and Galaxy instance were managed by [GBCS](https://www.embl.org/groups/genome-biology-computational-support/), [Bio-IT](https://bio-it.embl.de/), and [IT Services](https://www.embl.org/about/info/it-services/). Special thanks go to Charles Girardot and Jelle Scholtalbers.
